server/functional-test/src/test/java/org/zanata/feature/concurrentedit/ConcurrentAccessTest.kt

```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.stream.Collectors
+import javax.ws.rs.client.Entity
+import javax.ws.rs.client.Invocation
+import javax.ws.rs.core.MediaType
+
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder
+import org.junit.BeforeClass
+import org.junit.Test
+import org.junit.experimental.categories.Category
+import org.zanata.feature.Trace
+import org.zanata.feature.testharness.TestPlan.DetailedTest
+import org.zanata.feature.testharness.ZanataTestCase
+import org.zanata.rest.dto.resource.Resource
+import org.zanata.util.Constants
+import org.zanata.util.PropertiesHolder
+import org.zanata.util.ZanataRestCaller
+
+import org.assertj.core.api.Assertions.assertThat
+import org.zanata.util.ZanataRestCaller.buildSourceResource
+import org.zanata.util.ZanataRestCaller.buildTextFlow
+
+/**
+ * @author Patrick Huang [pahuang@redhat.com](mailto:pahuang@redhat.com)
+ */
+@Category(DetailedTest::class)
+class ConcurrentAccessTest : ZanataTestCase() {
+
+    @Trace(summary = "The system will handle concurrent document " +
+            "creation gracefully")
+    @Test(timeout = MAX_SHORT_TEST_DURATION.toLong())
+    @Throws(InterruptedException::class)
+    fun concurrentDocumentCreationWillNotCauseHibernateException() {
+        val projectSlug = "project"
+        val iterationSlug = "master"
+        val counter = AtomicInteger(1)
+        ZanataRestCaller().createProjectAndVersion(projectSlug,
+                iterationSlug, "gettext")
+
+        val threadCount = 5
+        val task = Callable {
+            val suffix = counter.getAndIncrement()
+            ZanataRestCaller().postSourceDocResource(
+                    projectSlug, iterationSlug, buildResource(suffix),
+                    false)
+        }
+        val tasks = Collections.nCopies(threadCount, task)
+        val executorService = Executors.newFixedThreadPool(threadCount)
+
+        val futures = executorService.invokeAll(tasks)
+
+        val result = getStatusCodes(futures)
+
+        val expectedReturnCode = Collections.nCopies(threadCount, 201)
+        assertThat(result).isEqualTo(expectedReturnCode)
+    }
+
+    companion object {
+
+        @BeforeClass
+        @Throws(Exception::class)
+        // Need to ensure that the correct concurrent slots are available
+        fun beforeClass() {
+            val path = "rest/configurations/c/max.concurrent.req.per.apikey"
+            val configRequest = clientRequestAsAdmin(path)
+            configRequest.put(Entity.json("6"))
+        }
+
+        private fun buildResource(suffix: Int): Resource {
+            return buildSourceResource("doc$suffix",
+                    buildTextFlow("res$suffix", "content$suffix"))
+        }
+
+        private fun getStatusCodes(futures: List<Future<Int>>): MutableList<Int> {
+            return futures.stream()
+                    .map { it -> getResult(it) }
+                    .collect(Collectors.toList())
```

Comment:
If this were new Kotlin code, I would write it like this:

```kt
        private fun getStatusCodes(futures: List<Future<Int>>): MutableList<Int> =
            futures.map { getResult(it) }.toMutableList()
```

Then there should be a quick-fix to use a method reference, probably something like this:

```kt
            futures.map { this::getResult) }.toMutableList()
```